textDocument/definition implementation for VS Code.

wpopielarski opened the pull request:
Trial of implementation of textDocument/definition request. Supports types only, and only in case when type is found in Zinc analysis. When source(s) is found then editor opens potential source(s). This simple implementation don't use semantic data (yet).

- [x] move `vscDefinition` task to separate file
- [x] add tests

eed3si9n committed:
implement window/logMessage

This sends sbt's log message as "window/logMessage" event to LSP.

wpopielarski committed:
Definition under construction.

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -1387,6 +1390,91 @@ object Defaults extends BuildCommon {
+  lazy val previousAnalysisParser = Def.setting {
+    Parsers.StringBasic
+  }
+  lazy val previousAnalysisCommand = Command.single("previousAnalysis") { (state, definition) =>
```
Why are we adding both an input key and a command here?

wpopielarski replied:
fixed

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -1387,6 +1390,91 @@ object Defaults extends BuildCommon {
+    val chnl = StandardMain.exchange.channels.collectFirst {
+      case c if c.name == cname => c
+    }
+    chnl.foreach(_.publishEvent(params, responseId))
```
If possible, we should use event logging here.

kiritsuku replied:
What is event logging?

eed3si9n replied:
Calling `infoEvent` on ManagedLogger, which then gets relayed to CommandExchange.

wpopielarski replied:
sorry, I don't get it.

eed3si9n commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -1387,6 +1390,91 @@ object Defaults extends BuildCommon {
+
+    val analysis = {
+      val compile = Keys.compile in Compile
+      Project.runTask(compile, updatedState, true).collect {
```
Is this invoking compilation task just to get to Analysis?

wpopielarski replied:
fixed, previousCompile only is used

wpopielarski commented:
still in progress, no worries about that. I have to clean it first.

2017-10-24 6:34 GMT+02:00 eugene yokota <notifications@github.com>:

> *@eed3si9n* requested changes on this pull request.
> ------------------------------
>
> In main/src/main/scala/sbt/Defaults.scala
> <https://github.com/sbt/sbt/pull/3660#discussion_r146451167>:
>
> > @@ -1387,6 +1390,91 @@ object Defaults extends BuildCommon {
>      }
>      analysisResult.analysis
>    }
> +  lazy val previousAnalysis = Def.inputKey[String]("gets previous analysis")
> +  lazy val previousAnalysisParser = Def.setting {
> +    Parsers.StringBasic
> +  }
> +  lazy val previousAnalysisCommand = Command.single("previousAnalysis") { (state, definition) =>
>
> Why are we adding both an input key and a command here?
> ------------------------------
>
> In main/src/main/scala/sbt/Defaults.scala
> <https://github.com/sbt/sbt/pull/3660#discussion_r146451248>:
>
> > +        }
> +      }
> +    val commandAnalysis = state.get(Keys.analysis)
> +    state.log.warn(s"command analysis: $commandAnalysis")
> +    val previousCompile = Project.runTask(Keys.previousCompile in Compile, updatedState).collect {
> +      case (newstate @ _, Value(v)) if v.analysis.isPresent => v.analysis.get
> +    }
> +    state.log.warn(s"previous analysis: $previousCompile")
> +    val cmdSrc = state.currentCommand.flatMap(_.source)
> +    import sbt.internal.langserver.{ Location, Position, Range }
> +    val params = Location(input.textDocument.uri, Range(Position(0, 0), Position(0, 10)))
> +    val cname = cmdSrc.map(_.channelName).get
> +    val chnl = StandardMain.exchange.channels.collectFirst {
> +      case c if c.name == cname => c
> +    }
> +    chnl.foreach(_.publishEvent(params, responseId))
>
> If possible, we should use event logging here.
> ------------------------------
>
> In main/src/main/scala/sbt/Defaults.scala
> <https://github.com/sbt/sbt/pull/3660#discussion_r146451437>:
>
> > +        }
> +        .toSeq
> +      state.log.warn(s"tasks $tasks")
> +      val complete = Aggregation.timedRun(updatedState, tasks, sbt.std.Transform.DummyTaskMap(Nil))
> +      state.log.warn("here")
> +      val r = complete.results.toEither.toOption.map { results =>
> +        state.log.warn("and there")
> +        results.map(_.value)
> +      }
> +      r
> +    }.getOrElse(Seq.empty)
> +    state.log.warn(s"aggregated analysis $aggrAnal")
> +
> +    val analysis = {
> +      val compile = Keys.compile in Compile
> +      Project.runTask(compile, updatedState, true).collect {
>
> Is this invoking compilation task just to get to Analysis?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#pullrequestreview-71393894>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzSzO9nk9CZ0Nx7FiY0AUZRWniGiN7ks5svWjugaJpZM4QB0Hy>
> .
>

wpopielarski committed:
Implementation task. In progress.

wpopielarski committed:
Basic functionality of vscode textDocument/definition.

wpopielarski committed:
Moving lsp definition request out of Defaults.

miklos-martin committed:
Fixes #3297 and #3531

Add commandName as an extension method in Command

miklos-martin committed:
add newline

wpopielarski committed:
Code rearrangement.

miklos-martin committed:
better naming + value class

wpopielarski committed:
More analysis from subprojects.

wpopielarski committed:
Merge branch '1.x' of https://github.com/sbt/sbt into vsc-definition

wpopielarski committed:
Merge branch '1.x' into vsc-definition

wpopielarski committed:
Adds copyrights header to file.

Final implementation.

Adds tests to Definition.

Adds reliable implementation of scala valid indentifier search (but with o(n^2) complexity).

saniyatech committed:
Removes reference to version 0.14.0 from a warning message

wpopielarski commented:
could anybody tell me what's wrong with tests? Locally they are passing here not.

miklos-martin committed:
move nameOption to the trait

wpopielarski reviewed (commented):
ready to be checked again

laughedelic commented:
Hi @wpopielarski! I tried to test this branch, but couldn't make it work.

1. There is one thing that I had to change locally: to comment these two lines
  https://github.com/sbt/sbt/blob/9232a8fbc872c2f94a91ec82d797a11140a271a0/main/src/main/scala/sbt/Defaults.scala#L1412-L1413  
    This is called in the `compileIncremental` task, that I guess, is used in your `lspDefinition` in some way. The result is that servers bombards client with _empty_ `publishDiagnostics` notifications (to clean up), but this happens in response to **every** `textDocument/definition` request. On the sbt codebase it means that every request takes at least 20 seconds. And even on a small project it's ~5s.

    Commenting these lines out is not a solution, but the point is, that `textDocument/definition` shouldn't trigger diagnostics cleanup at all

    On a sidenote, I think that diagnostics cleanup should happen somewhere else. An alternative approach is to store diagnostics and every time new ones are generated, "merge" them and send only the necessary changes, this way there is no need to send lots of empty diagnostics on every compile. This is just an idea and, of course, out of scope of this PR.

2. After that I could see `textDocument/definition` requests/responses. And I couldn't get any results: responses are _always empty_. I tried on the sbt code, on some work project, on a tiny test project — it didn't work anywhere. Could you provide an example code on which it is supposed to work well?

By the way, I tested it with Atom, but I think that there is nothing VSCode-specific happening here, so it sould work the same in both editors. I should install VSCode and try it out too...

laughedelic commented:
No luck in VS Code either:
<img width="447" alt="screen shot 2017-11-05 at 05 39 16" src="https://user-images.githubusercontent.com/766656/32412064-46f1254c-c1ec-11e7-8e64-fdc80943789d.png">

I have a feeling that I'm holding it from the wrong end.. 😅

wpopielarski commented:
Just Foo is not class, trait, object so not available in zinc analysis
file. Right now jump to definition is made from these analysis files not
scalameta-sbt plugin.

2017-11-05 5:44 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> No luck in VS Code either:
> [image: screen shot 2017-11-05 at 05 39 16]
> <https://user-images.githubusercontent.com/766656/32412064-46f1254c-c1ec-11e7-8e64-fdc80943789d.png>
>
> I have a feeling that I'm holding it from the wrong end.. 😅
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341948887>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzS9iYfaz6e_pMAofAq5_ViyLb-4PKks5szT1KgaJpZM4QB0Hy>
> .
>

wpopielarski commented:
Alexey,

there are not my lines and I think you should ask Eugen Yokota about the
reason of these 2 lines. Actually it is hard to say why you get empty
response. One of the reason can be the lack of zinc analysis cache files
which are used to find products of compilation. So probably you have to
compile project first and than try to jump to definition. And because type
identifiers are got from these files so if you are asking about identifier
which is not found there (even if it is scala Int, String or any other
common type) you get 'No definition found'.

2017-11-05 5:25 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> Hi @wpopielarski <https://github.com/wpopielarski>! I tried to test this
> branch, but couldn't make it work.
>
>    1.
>
>    There is one thing that I had to change locally: to comment these two
>    lines
>    https://github.com/sbt/sbt/blob/9232a8fbc872c2f94a91ec82d797a1
>    1140a271a0/main/src/main/scala/sbt/Defaults.scala#L1412-L1413
>    <https://github.com/sbt/sbt/blob/9232a8fbc872c2f94a91ec82d797a11140a271a0/main/src/main/scala/sbt/Defaults.scala#L1412-L1413>
>    This is called in the compileIncremental task, that I guess, is used
>    in your lspDefinition in some way. The result is that servers bombards
>    client with *empty* publishDiagnostics notifications (to clean up),
>    but this happens in response to *every* textDocument/definition
>    request. On the sbt codebase it means that every request takes at least 20
>    seconds. And even on a small project it's ~5s.
>
>    Commenting these lines out is not a solution, but the point is, that
>    textDocument/definition shouldn't trigger diagnostics cleanup at all
>
>    On a sidenote, I think that diagnostics cleanup should happen
>    somewhere else. An alternative approach is to store diagnostics and every
>    time new ones are generated, "merge" them and send only the necessary
>    changes, this way there is no need to send lots of empty diagnostics on
>    every compile. This is just an idea and, of course, out of scope of this PR.
>    2.
>
>    After that I could see textDocument/definition requests/responses. And
>    I couldn't get any results: responses are *always empty*. I tried on
>    the sbt code, on some work project, on a tiny test project — it didn't work
>    anywhere. Could you provide an example code on which it is supposed to work
>    well?
>
> By the way, I tested it with Atom, but I think that there is nothing
> VSCode-specific happening here, so it sould work the same in both editors.
> I should install VSCode and try it out too...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341948327>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzS2pFqzTNw0p62upb8CX84OqTz1Hvks5szTjLgaJpZM4QB0Hy>
> .
>

laughedelic commented:
> Just Foo is not class, trait, object so not available in zinc analysis file.

> One of the reason can be the lack of zinc analysis cache files
> which are used to find products of compilation. So probably you have to
> compile project first and than try to jump to definition. 

<img width="367" alt="screen shot 2017-11-05 at 17 17 33" src="https://user-images.githubusercontent.com/766656/32416681-6ef3d4c0-c24d-11e7-8ee2-1b925e20684c.png">

I definitely compiled the project and I don't get definitions on _any symbol_ in this little piece of code. Tried with defining a class, a trait, an object — nothing works..
As I said before, could you provide some example code which works for you?

> there are not my lines and I think you should ask Eugen Yokota about the reason of these 2 lines.

Sure, I wrote that more for @eed3si9n 

> Right now jump to definition is made from these analysis files not scalameta-sbt plugin.

I was actually thinking about it recently and was going to ask you. It seems that using information from the scalameta's semantic DB would be a straightforward way to get any symbol information for the language server. But I guess it's out of sbt's scope, because it requires a third-party plugin.. 
Anyway, in this light, is it still worth digging zinc analysis cache to get only partial information only for some symbols? Or would it make more sense to just implement this functionality outside of sbt, using scalameta-sbt (and interact with it through sbt server)?

wpopielarski commented:
Really hard to say. It's working for me. So my question is: do you get
compilation errors? This part should work as well. You should get also info
as logs about received requests by sbt server

2017-11-05 17:32 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> Just Foo is not class, trait, object so not available in zinc analysis
> file.
>
> One of the reason can be the lack of zinc analysis cache files
> which are used to find products of compilation. So probably you have to
> compile project first and than try to jump to definition.
>
> [image: screen shot 2017-11-05 at 17 17 33]
> <https://user-images.githubusercontent.com/766656/32416681-6ef3d4c0-c24d-11e7-8ee2-1b925e20684c.png>
>
> I definitely compiled the project and I don't get definitions on *any
> symbol* in this little piece of code. Tried with defining a class, a
> trait, an object — nothing works..
> As I said before, could you provide some example code which works for you?
>
> there are not my lines and I think you should ask Eugen Yokota about the
> reason of these 2 lines.
>
> Sure, I wrote that more for @eed3si9n <https://github.com/eed3si9n>
>
> Right now jump to definition is made from these analysis files not
> scalameta-sbt plugin.
>
> I was actually thinking about it recently and was going to ask you. It
> seems that using information from the scalameta's semantic DB would be a
> straightforward way to get any symbol information for the language server.
> But I guess it's out of sbt's scope, because it requires a third-party
> plugin..
> Anyway, in this light, is it still worth digging zinc analysis cache to
> get only partial information only for some symbols? Or would it make more
> sense to just implement this functionality outside of sbt, using
> scalameta-sbt (and interact with it through sbt server)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341985989>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzS1AHIamWirF9gIfKVQoqwuotplucks5szeMegaJpZM4QB0Hy>
> .
>

laughedelic commented:
> So my question is: do you get
> compilation errors? This part should work as well. You should get also info
> as logs about received requests by sbt server

It compiles well, no errors (really simple code on the screenshot above). And I do see logs about receiving requests. I also see empty responses in Atom (can I see client-server communication log somewhere in VS Code?).

wpopielarski commented:
about scalameta-sbt. Finding a symbol helps to find its definition by
determining its type or some local stable identifier or whatever else but
it still need work to figure out its source file and position in this file.
So analysis files are just first step in this way I guess

2017-11-05 18:00 GMT+01:00 Wieslaw Popielarski <wpopielarski@gmail.com>:

> Really hard to say. It's working for me. So my question is: do you get
> compilation errors? This part should work as well. You should get also info
> as logs about received requests by sbt server
>
> 2017-11-05 17:32 GMT+01:00 Alexey Alekhin <notifications@github.com>:
>
>> Just Foo is not class, trait, object so not available in zinc analysis
>> file.
>>
>> One of the reason can be the lack of zinc analysis cache files
>> which are used to find products of compilation. So probably you have to
>> compile project first and than try to jump to definition.
>>
>> [image: screen shot 2017-11-05 at 17 17 33]
>> <https://user-images.githubusercontent.com/766656/32416681-6ef3d4c0-c24d-11e7-8ee2-1b925e20684c.png>
>>
>> I definitely compiled the project and I don't get definitions on *any
>> symbol* in this little piece of code. Tried with defining a class, a
>> trait, an object — nothing works..
>> As I said before, could you provide some example code which works for you?
>>
>> there are not my lines and I think you should ask Eugen Yokota about the
>> reason of these 2 lines.
>>
>> Sure, I wrote that more for @eed3si9n <https://github.com/eed3si9n>
>>
>> Right now jump to definition is made from these analysis files not
>> scalameta-sbt plugin.
>>
>> I was actually thinking about it recently and was going to ask you. It
>> seems that using information from the scalameta's semantic DB would be a
>> straightforward way to get any symbol information for the language server.
>> But I guess it's out of sbt's scope, because it requires a third-party
>> plugin..
>> Anyway, in this light, is it still worth digging zinc analysis cache to
>> get only partial information only for some symbols? Or would it make more
>> sense to just implement this functionality outside of sbt, using
>> scalameta-sbt (and interact with it through sbt server)?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/sbt/sbt/pull/3660#issuecomment-341985989>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ALEzS1AHIamWirF9gIfKVQoqwuotplucks5szeMegaJpZM4QB0Hy>
>> .
>>
>
>

wpopielarski commented:
I don't mean 'correct code' :) I mean if you make failing snippet do you
get error compilation back?

2017-11-05 18:05 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> So my question is: do you get
> compilation errors? This part should work as well. You should get also info
> as logs about received requests by sbt server
>
> It compiles well, no errors (really simple code on the screenshot above).
> And I do see logs about receiving requests. I also see empty responses in
> Atom (can I see client-server communication log somewhere in VS Code?).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341988324>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzS6M1SeAPIeiS4uwlqDfIk9unFDX_ks5szerbgaJpZM4QB0Hy>
> .
>

laughedelic commented:
> I don't mean 'correct code' :) I mean if you make failing snippet do you get error compilation back?

😄 I see. Yes, this part works fine, errors are reported as diagnostics.

> Finding a symbol helps to find its definition by determining its type or some local stable identifier or whatever else but it still need work to figure out its source file and position in this file. So analysis files are just first step in this way I guess

You mean that even with scalameta we will need analysis cache from zinc to map symbols to the source locations? I thought it's also stored in the SemanticDB, but now I see that it's not.

wpopielarski commented:
but what you can see in log from:

https://github.com/sbt/sbt/pull/3660/files#diff-82bb5e5687c35f3190dd111f8a9d93bbR166

2017-11-05 18:18 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> I don't mean 'correct code' :) I mean if you make failing snippet do you
> get error compilation back?
>
> 😄 I see. Yes, this part works fine, errors are reported as diagnostics.
>
> Finding a symbol helps to find its definition by determining its type or
> some local stable identifier or whatever else but it still need work to
> figure out its source file and position in this file. So analysis files are
> just first step in this way I guess
>
> You mean that even with scalameta we will need analysis cache from zinc to
> map symbols to the source locations? I thought it's also stored in the
> SemanticDB, but now I see that it's not.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341989270>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzSwYGM5YTVMR0iqI9uK8F929c2Hj6ks5sze4DgaJpZM4QB0Hy>
> .
>

laughedelic commented:
Oh, I didn't see these logs. Should they appear in the sbt shell? If I don't get them, what am I doing wrong?

wpopielarski commented:
even this one?
https://github.com/sbt/sbt/pull/3660/files#diff-82bb5e5687c35f3190dd111f8a9d93bbR134

2017-11-05 18:39 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> Oh, I didn't see these logs. Should they appear in the sbt shell? If I
> don't get them, what am I doing wrong?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341990799>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzSy9Uq_cDp4_99JrVVPDo6O8Dxiy6ks5szfLXgaJpZM4QB0Hy>
> .
>

wpopielarski commented:
so as I got you, you checked out vsc-definition branch and published it
locally and in your test project you set sbt version to this locally
published and it failed?

2017-11-05 18:43 GMT+01:00 Wieslaw Popielarski <wpopielarski@gmail.com>:

> even this one? https://github.com/sbt/sbt/pull/3660/files#diff-
> 82bb5e5687c35f3190dd111f8a9d93bbR134
>
> 2017-11-05 18:39 GMT+01:00 Alexey Alekhin <notifications@github.com>:
>
>> Oh, I didn't see these logs. Should they appear in the sbt shell? If I
>> don't get them, what am I doing wrong?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/sbt/sbt/pull/3660#issuecomment-341990799>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ALEzSy9Uq_cDp4_99JrVVPDo6O8Dxiy6ks5szfLXgaJpZM4QB0Hy>
>> .
>>
>
>

laughedelic commented:
Nope, I only see (in the console)

```
[debug] onRequestMessage: JsonRpcRequestMessage(2.0, 8, textDocument/definition, {"textDocument":{"uri":".../test.scala"},"position":{"line":13,"character":17}}})
```

And then empty `textDocument/publishDiagnostics` notifications. I also see that Atom receives an empty response:

```
Scala (sbt) rpc.sendRequest textDocument/definition sending Object ...
...
Scala (sbt) rpc.sendRequest textDocument/definition received (5673ms) []
```

But no logs from `lspDefinitionTask` anywhere.

laughedelic commented:
Yes, I checked your branch, published locally and I run test project with that locally published version. And I know that it's that version because I see the changes and I see logs about definition requests. And I also see `lspDefinition` task in the test project sbt shell. Btw, how can I test it directly from the shell?

wpopielarski commented:
can you ran task 'lspDefinition` in sbt console?

2017-11-05 18:51 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> Yes, I checked your branch, published locally and I run test project with
> that locally published version. And I know that it's that version because I
> see the changes and I see logs about definition requests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341991685>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzSwiA5JePeuZ2yBQQ64sM1N-ti89Bks5szfWrgaJpZM4QB0Hy>
> .
>

laughedelic commented:
Yes, but I don't know what arguments to give it. I tried with the raw JSON request, but it didn't work:

```
 lspDefinition """{ "jsonrpc": "2.0", "method": "textDocument/definition", "params": {"textDocument":{"uri":"file:///.../test.scala"},"position":{"line":15,"character":13}}} }"""
[warn] Incorrect definition request Some({ "jsonrpc": "2.0", "method": "textDocument/definition", "params": {"textDocument":{"uri":"file:///.../test.scala"},"position":{"line":15,"character":13}}} })
[success] Total time: 0 s, completed Nov 5, 2017 6:54:38 PM
```

wpopielarski commented:
ok, first 'lspDefinition {"textDocument":{"uri":"file:///.../test.
scala"},"position":{"line":15,"character":13}' but it looks like this task
is not visible, i don't know... I take a look tomorrow at this

2017-11-05 19:08 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> Yes, but I don't know what arguments to give it. I tried with the raw JSON
> request, but it didn't work:
>
>  lspDefinition """{ "jsonrpc": "2.0", "method": "textDocument/definition", "params": {"textDocument":{"uri":"file:///.../test.scala"},"position":{"line":15,"character":13}}} }"""
> [warn] Incorrect definition request Some({ "jsonrpc": "2.0", "method": "textDocument/definition", "params": {"textDocument":{"uri":"file:///.../test.scala"},"position":{"line":15,"character":13}}} })
> [success] Total time: 0 s, completed Nov 5, 2017 6:54:38 PM
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341992807>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzS_oXYzy5qZzCievOLMeUXem7zqdqks5szfmfgaJpZM4QB0Hy>
> .
>

laughedelic commented:
OK, I figured out that it was about log level, I changed it to `warn` in the code and now I see the logs:

```
> lspDefinition {"textDocument":{"uri":"file:///.../test.scala"},"position":{"line":15,"character":13}}
[warn] lsp-definition raw request: List({"textDocument":{"uri":"file:///.../test.scala"},"position":{"line":15,"character":13}})
[warn] lsp-definition found line:   def qux(f: Foo): String = f.b
[debug] langNotify: JsonRpcNotificationMessage(2.0, textDocument/publishDiagnostics, {"uri":"file:/.../test.scala","diagnostics":[]})
[warn] lsp-definition potentials: Set()
[warn] lsp-definition potentials: Set()
[success] Total time: 1 s, completed Nov 5, 2017 7:27:56 PM
```

wpopielarski commented:
but `Foo` is `type Foo` or class Foo? Sorry I scolled up and see it, it's a class. so what you get when call `show previousCompile` ? Ok eureka my bug my fault. You defined `Foo` in default package and unfortunately I don't look for classes,traits, objects there. Need to fix it. But when you started to play with it use packages please, and let me know. you earned medal for testers 👍

laughedelic commented:
Class `Foo` just as you see on the screenshot from https://github.com/sbt/sbt/pull/3660#issuecomment-341985989, no other `Foo` in scope:

```scala
class Foo(val a: Int, val b: String)

object Buh {
  val foo: Foo = new Foo(1, "a")

  def qux(f: Foo): String = f.b

  qux(foo)
}
```

This is the only code in the one file in the project.

wpopielarski commented:
see at:
https://github.com/sbt/sbt/pull/3660/files#diff-82bb5e5687c35f3190dd111f8a9d93bbR61
I need some integration test too I see

laughedelic commented:
Didn't notice your edit in https://github.com/sbt/sbt/pull/3660#issuecomment-341994983 at first.. 😄 

I added a package, tried putting `Foo` inside the `Buh` object, nothing helps.. Here's `previousCompile` output:

```
> show previousCompile
[info] PreviousResult(analysis: Optional[Analysis: 3 Scala sources, 5 classes, 2 binary dependencies], setup: Optional[MiniSetup(output: SingleOutput(/Users/laughedelic/dev/laughedelic/scala-adder/target/scala-2.12/classes), options: MiniOptions(classpathHash: [Lxsbti.compile.FileHash;@68d57037, scalacOptions: [Ljava.lang.String;@5f0669d9, javacOptions: [Ljava.lang.String;@1c4f3b3d), compilerVersion: 2.12.4, order: Mixed, storeApis: true, extra: [Lxsbti.T2;@211d17c)])
[success] Total time: 0 s, completed Nov 5, 2017 8:04:14 PM
```

wpopielarski commented:
generally I tested it on the whole `sbt` project itself and working quite cool (I even did demo for @eed3si9n  and @sschaef ). Will see tomorrow then... Sorry for that.

wpopielarski commented:
good news is that is is working for simple projects too. But I have some
complex configuration taken from sbt project. I try to narrow it.

2017-11-05 20:06 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> Didn't notice your edit in #3660 (comment)
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341994983> at first..
> 😄
>
> I added a package, tried putting Foo inside the Buh object, nothing
> helps.. Here's previousCompile output:
>
> > show previousCompile
> [info] PreviousResult(analysis: Optional[Analysis: 3 Scala sources, 5 classes, 2 binary dependencies], setup: Optional[MiniSetup(output: SingleOutput(/Users/laughedelic/dev/laughedelic/scala-adder/target/scala-2.12/classes), options: MiniOptions(classpathHash: [Lxsbti.compile.FileHash;@68d57037, scalacOptions: [Ljava.lang.String;@5f0669d9, javacOptions: [Ljava.lang.String;@1c4f3b3d), compilerVersion: 2.12.4, order: Mixed, storeApis: true, extra: [Lxsbti.T2;@211d17c)])
> [success] Total time: 0 s, completed Nov 5, 2017 8:04:14 PM
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-341996737>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzSzSDjMWDfSGekatiNlfS6tUdx3Mfks5szgdGgaJpZM4QB0Hy>
> .
>

wpopielarski commented:
well, looks that extremely simple build.sbt must work. The only important
observation is that the very root project folder must be open by vs
code.Like this:

![screenshot from 2017-11-06 12 56 20](https://user-images.githubusercontent.com/11613003/32441450-09db414e-c2f8-11e7-8dca-992ff9c605f3.png)

wpopielarski commented:
and one remark: the selected definition is not found in the very first time for clean project (no compilation done). The second attempt is successful.
REMARK: I didn't test it without removed by you lines in `Defaults`

wpopielarski commented:
ok, I see big problems when these two lines are removed. What is interesting, I cannot drop back to previous behavior.

wpopielarski committed:
Fixes problem with finding valid scala identifier.

Adds better search for valid scala symbol.

Adds in-memory cache to sbt and lspCollectAnalyses task.

Adds lspCollectAnalyses to compile and initialize LSP actions.

laughedelic commented:
What do you mean? I already uncommented them and I get those empty diagnostics again. But I still don't get definitions. Can you share your acme test project?

wpopielarski commented:
funny thing, I can see definition of `Foo` in `internal.Foo.scala` only
![screenshot from 2017-11-06 15 15 48](https://user-images.githubusercontent.com/11613003/32445366-ceccac88-c305-11e7-8c8a-a2874c158cd8.png)

wpopielarski commented:
so I don't have any idea how `previousCompile` is working
![screenshot from 2017-11-06 15 19 49](https://user-images.githubusercontent.com/11613003/32445433-f62c3596-c305-11e7-8c04-1c3df8ac3c4b.png)

laughedelic commented:
🙁 it's confusing..

On a sidenote, guys from scalameta are actively working on the go-to-definition implementation right now: https://github.com/scalameta/language-server/pull/8

wpopielarski commented:
to calm down in 'previousCompile` usage. It's ok. I suffer with toolbox problem. testing...

wpopielarski commented:
@laughedelic could you please, get the commit up and check if definition is now working? There was my fault. Unfortunately `ToolBox.parse` is more powerful I expected. So now to check if given token is longest valid scala identifier I ask tb to compile this snippet of code.

laughedelic commented:
Done 👌 It worked for me for the first time! But only with objects. Doesn't work with anything else..

<img width="667" alt="screen shot 2017-11-06 at 19 06 56" src="https://user-images.githubusercontent.com/766656/32456403-c1e6bfe8-c325-11e7-88f9-9a2ba084634d.png">

(UPD: better screenshot, so you can see both files content and the project structure)

I also noticed that response contains same reference object twice.

wpopielarski commented:
really, not working for plain classes? or classes in objects?

2017-11-06 19:05 GMT+01:00 Alexey Alekhin <notifications@github.com>:

> Done 👌 It worked for me for the first time! But only with objects.
> Doesn't work with anything else..
>
> [image: screen shot 2017-11-06 at 19 04 34]
> <https://user-images.githubusercontent.com/766656/32456294-67ad8746-c325-11e7-9c6f-45de0f785654.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-342234125>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzSxh19seOYooXlLXw_ETpuVf0g9v4ks5sz0p4gaJpZM4QB0Hy>
> .
>

laughedelic commented:
Nope. On the screenshot above I have class `Foo`, tried putting it in or out of the object, doesn't work either way.

wpopielarski commented:
I'm just asking because right now is very simple (and naive) assumption
about simple class names. Only suffices .A, .A$, $A$ and $A are available. But this naive assumption is good enough besides default package

2017-11-06 19:24 GMT+01:00 Wieslaw Popielarski <wpopielarski@gmail.com>:

> really, not working for plain classes? or classes in objects?
>
> 2017-11-06 19:05 GMT+01:00 Alexey Alekhin <notifications@github.com>:
>
>> Done 👌 It worked for me for the first time! But only with objects.
>> Doesn't work with anything else..
>>
>> [image: screen shot 2017-11-06 at 19 04 34]
>> <https://user-images.githubusercontent.com/766656/32456294-67ad8746-c325-11e7-9c6f-45de0f785654.png>
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/sbt/sbt/pull/3660#issuecomment-342234125>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ALEzSxh19seOYooXlLXw_ETpuVf0g9v4ks5sz0p4gaJpZM4QB0Hy>
>> .
>>
>
>

wpopielarski commented:
I don't know about classes.... Here you are:
![screenshot from 2017-11-07 09 29 53](https://user-images.githubusercontent.com/11613003/32484033-c4e920e0-c39e-11e7-8d70-6ee6fea46dfb.png)

wpopielarski commented:
@laughedelic could you play a while?
![screenshot from 2017-11-07 16 56 11](https://user-images.githubusercontent.com/11613003/32503093-a21657c0-c3dc-11e7-9f17-e44ee36b06f4.png)

laughedelic commented:
@wpopielarski I'll try it out tonight 👍

laughedelic commented:
@wpopielarski Now it **does** work for objects, classes and traits 👍

eed3si9n commented on main/src/main/scala/sbt/internal/server/LanguageServerProtocol.scala:
```diff
@@ -60,6 +58,12 @@ private[sbt] trait LanguageServerProtocol extends CommandChannel {
+      case "textDocument/definition" =>
+        import sjsonnew.support.scalajson.unsafe.CompactPrinter
+        append(
+          Exec(s"lspDefinition ${CompactPrinter(json)}",
```
The query for definition, if possible, should happen concurrently with whatever that's happening in the command processing engine. For example, the user could be in the middle of `test` that takes 10 min. I think `textDocument/definition` should return while the test is running, not afterwards. So I don't think we should use task or input task here.

wpopielarski replied:
looks like fixed I hope

wpopielarski commented:
so should be command used here? This part of code is using
`previousCompile` to collect analyses. Is any other way to figure out
analysis files names and locations?

2017-11-08 5:05 GMT+01:00 eugene yokota <notifications@github.com>:

> *@eed3si9n* requested changes on this pull request.
> ------------------------------
>
> In main/src/main/scala/sbt/internal/server/LanguageServerProtocol.scala
> <https://github.com/sbt/sbt/pull/3660#discussion_r149570600>:
>
> > @@ -60,6 +58,12 @@ private[sbt] trait LanguageServerProtocol extends CommandChannel {
>          langRespond(InitializeResult(serverCapabilities), Option(request.id))
>        case "textDocument/didSave" =>
>          append(Exec("compile", Some(request.id), Some(CommandSource(name))))
> +      case "textDocument/definition" =>
> +        import sjsonnew.support.scalajson.unsafe.CompactPrinter
> +        append(
> +          Exec(s"lspDefinition ${CompactPrinter(json)}",
>
> The query for definition, if possible, should happen concurrently with
> whatever that's happening in the command processing engine. For example,
> the user could be in the middle of test that takes 10 min. I think
> textDocument/definition should return while the test is running, not
> afterwards. So I don't think we should use task or input task here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#pullrequestreview-74985143>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzS8noLoCFYOGrqAIe_E5uLMqm4Z3Nks5s0SiMgaJpZM4QB0Hy>
> .
>

eed3si9n commented:
I am guessing that you'd need to set up a cache of Analysis files yourself based on the known locations.

wpopielarski commented:
that's clear. thanks

2017-11-08 6:45 GMT+01:00 eugene yokota <notifications@github.com>:

> I am guessing that you'd need to set up a cache of Analysis files yourself
> based on the known locations.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sbt/sbt/pull/3660#issuecomment-342717057>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALEzS0gmkQV7NY8LIU4EYkegGrxuEUh6ks5s0T_0gaJpZM4QB0Hy>
> .
>

wpopielarski committed:
Async LSP Definition implementation.

Groups imports and corrects logging info.

Optimizes access to analyses files.

Parallel computation should speed up processing.

Adds super efficient optimization.

Encodes class names.

Code style correction.

Escapes regexp characters in class name.

wpopielarski commented:
@laughedelic @sschaef  @eed3si9n could you play with is a little please?

laughedelic commented:
Hi @wpopielarski! I just tried out 9152ece. It's much more stable than before! It works on all objects, classes and traits in my test project 👍 

<img width="488" alt="screen shot 2017-11-14 at 04 57 40" src="https://user-images.githubusercontent.com/766656/32762300-fc6e498c-c8f8-11e7-8a51-61dc489c715b.png">

Although I can't say that it's usable yet. There are several problems:
* When I hover on empty space it responses with a lot of `{code: -32700, message: "Symbol not found in definition request"}` errors. I think it shouldn't send an error and `ParseError` is probably meant for the cases when you cannot parse the request itself. I also checked what other language servers do: 
  - Flow server sends a response with `null` result (indicating success) and 
  - and Python sends an empty response `[]` (no symbols found).
  IMO Flow server does the right thing
* It's unclear what to expect from this functionality, which things are supposed should work and which not. I think you should clarify it. You say 
  > Supports types only, and only in case when type is found in Zinc analysis

  but I have no idea what's found in Zinc analysis. Also, does it mean that it will never work for values/methods/etc.?
* It's super slow 🐌  Some unsuccessful(!) requests take ~1.5s, and most successul ones take >1s. I see a spinning wheel on my mouse pointer 😅 I understand that this is work in progress and something could be optimized, but I think it's supposed to be several orders of magnitude faster.. 

----

Overall, I think that this is a very good experiment, but I really doubt now that it should be a part of the sbt server core. Taking into account that it's implemented through a task, it could probably be an sbt plugin? Or is it really tied to the sbt internals?

In my opinion sbt server development should be focused more on the interaction with the build, not sources analysis. There are already two (three if you count Dotty) language servers for Scala that provide this functionality and many other things: ENSIME and Scalameta (see https://github.com/scalameta/language-server/pull/23). 

@wpopielarski I don't want to discourage you from improving this, I just want everybody working on things related to Scala and LSP be more aware of each other's work.

wpopielarski commented:
@laughedelic class trait and object are for good start. zinc can give info about methods and class fields too. I think we will use scalameta finally. meantime I add super efficient optimization (if working :) )

laughedelic committed:
Publish ExecStatusEvent taking into account its channel name

laughedelic committed:
Log StringEvents and MessageEvents to the LSP client

laughedelic committed:
Tweaked MessageType.fromLevelString conversion

wpopielarski commented:
well, even with simple classes is problem because names are encoded so there is no other possibility than use scalameta :). But is NameTransformer

Duhemm committed:
Remove 2 useless files in protocol

They have been moved to Contraband since, but never removed.

xuwei-k committed:
fix #3591. "Not a valid key: console-quick"

xuwei-k committed:
make strict some methods in Completions

xuwei-k committed:
update specs2 4.0.1

wpopielarski committed:
Adds backticks to class/trait/object name.

Adapts tests to changed specs2 dependency.

Tiny fixes.

Removes Scala IDE compiler clues.

dwijnand commented on main/src/main/scala/sbt/lsp/Definition.scala:
```diff
@@ -0,0 +1,330 @@
+import java.nio.file.{ Files, Paths }
+import sbt.StandardMain
+
+object Definition {
```
Either this object should be `private[sbt]` or the whole `lsp` package should be at `sbt.internal.lsp`.

dwijnand reviewed (changes requested):
I left one comment about making sure that private/internal things are private.

Additional things for this PR:
1. squash the commits
2. remove the weird, side diffs from this PR (e.g `import Configurations.Test`, etc)
3. check and report what the transitive dependencies of scalacache-caffeine are.

wpopielarski commented:
@dwijnand 
looks like there is no dependency in compile. For simple maven project dependency:tree returns:
```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ my-app ---
[INFO] com.mycompany.app:my-app:jar:1.0-SNAPSHOT
[INFO] \- com.github.ben-manes.caffeine:caffeine:jar:2.6.0:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

dwijnand commented:
the transitive dependencies of scalacache-caffeine look fine too:

```
com.github.cb372:scalacache-caffeine_2.12:0.20.0 [S]
+-com.github.ben-manes.caffeine:caffeine:2.5.6
+-com.github.cb372:scalacache-core_2.12:0.20.0 [S]
| +-org.scala-lang:scala-reflect:2.12.3 [S]
| +-org.slf4j:slf4j-api:1.7.25
|
+-org.slf4j:slf4j-api:1.7.25
```

wpopielarski committed:
Merge branch '1.x' into vsc-definition

eed3si9n commented:
I'm closing this in favor of https://github.com/sbt/sbt/pull/3777

eed3si9n commented:
If you could give me push rights, I could put those commits here too.

wpopielarski commented:
@eed3si9n thanks a lot for flattening history. I did the mess. And I think if https://github.com/sbt/sbt/pull/3777 is merged so is there a need to commit it to this branch? If I got you correctly.

eed3si9n commented:
@wpopielarski I guess not. Thanks for the contribution :)
